mlflow/dspy/callback.py

```diff
@@ -204,6 +215,80 @@ def on_tool_end(
         if call_id in self._call_id_to_span:
             self._end_span(call_id, outputs, exception)
 
+    def on_evaluate_start(self, call_id: str, instance: Any, inputs: dict[str, Any]):
+        """
+        Callback handler at the beginning of evaluation call. Available with DSPy>=2.6.9.
+        This callback starts a nested run for each evaluation call if called within optimization,
+        and created a new run when called outside of optimization and there is no active run.
+        """
+        if not get_autologging_config(FLAVOR_NAME, "log_evals"):
+            return
+        if devset := inputs.get("devset"):
+            is_minibatch = len(devset) < len(instance.devset)
+        else:
+            is_minibatch = False
+
+        if self.optimizer_stack_level > 0:
+            key = "minibatch" if is_minibatch else "full"
+            step = self._evaluation_counter[key]
+            run = mlflow.start_run(run_name=f"eval_{key}_{step}", nested=True)
+            self._evaluation_counter[key] += 1
+            self._call_id_to_eval_state[call_id] = (is_minibatch, step)
+            self._call_id_to_run_id[call_id] = run.info.run_id
+        else:
+            if mlflow.active_run() is None:
+                run = mlflow.start_run()
+                self._call_id_to_run_id[call_id] = run.info.run_id
+        if program := inputs.get("program"):
```

Comment:
We should probably log some params to the run so that users know what they are looking at:

- For mipro - that's the combo of instruction + demo set for each predict.
- For BootstrapFS, that's the demo set.
- For BetterTogether, that's the finetuned model path/endpoint + normal state.

Reply:
Good point. I first thought to use log_params, but unfortunately, it only accepts {"key": "value} style. So logging model state as an artifact is a promising direction. We can try to show some params as follow up.

Reply:
probably:
`log_param("instruction_candidates", ...)` and `log_param("demo_candidates", ...)`?

Reply:
As we discussed, will follow up on this in the next PR.